tools/syz-testbed: produce results indefinitely

a-nogikh opened the pull request:
Creation of instances only at the start of the tool limits the scale of
the experiments.

Instead of setting up instances at the beginning, create, start and
stop them dynamically. This allows to conduct much larger experiments
while keeping the maximal resource consumption under control. When an
instance has been executing for the specified time, it is stopped and
replaced by another one.

Introduce two new parameters that control it:
- `max_instance` - the maximal number of simultaneously running instances.
- `run_hours` - the lifetime (in hours) of a single instance.

Now the longer the tool runs, the more test results will be accumulated
for each checkout specified in the configuration file.

Also, add a simple web interface, which is controlled by two optional
parameters:
- `http` - host and port on which the tool should listen.
- `benchcmp` - the path to the syz-benchcmp executable.

dvyukov commented on tools/syz-testbed/html.go:
```diff
@@ -0,0 +1,186 @@
+		return
+	}
+
+	// TODO: move syz-benchcmp functionality to pkg/ and just import it?
```
+1 that's better on all fronts

dvyukov commented on tools/syz-testbed/testbed.go:
```diff
@@ -82,150 +78,220 @@ func main() {
 	}
 	if managerCfg.HTTP == "" {
-		managerCfg.HTTP = ":50000"
+		managerCfg.HTTP = "0.0.0.0:0"
```
You also removed increasePort, will it now work with anything other than 0 port? Or users should only override the IP address but keep 0 port?

a-nogikh replied:
Manual port assignment turned out to be somewhat troublesome. We would also have to set RPC ports - otherwise randomly taken free ports sometimes overlap with the ports that were assigned to syz-managers. I've seen this happening a lot in the currently upstreamed version of the tool, and now that the tool periodically restarts the managers, this would happen more and more often.

Also, the experience has shown that there's actually no need to access individual dashboards - therefore we don't have to care about the ports at all.

Regarding this particular changed line - probably it'd be better just to always override HTTP with ":0".

dvyukov replied:
Some users may want to use "localhost:0", so I think it's fine as it is now.

dvyukov commented on tools/syz-testbed/testbed.go:
```diff
@@ -82,150 +78,220 @@ func main() {
-	for _, co := range checkouts {
-		for _, instance := range co.Instances {
-			go runInstance(instance)
+	if cfg.HTTP != "" {
```
I think it's better to provide ":0" default and always start the http server.

a-nogikh replied:
But what would be the use of a server which is bound to some random and unknown port?

dvyukov replied:
We would need to print the actual port to the console.

dvyukov commented on tools/syz-testbed/testbed.go:
```diff
@@ -82,150 +78,220 @@ func main() {
+	for _, checkout := range ctx.Checkouts {
+		until := "-"
+		if ctx.NextRestart.After(time.Now()) {
+			until = time.Until(ctx.NextRestart).String()
```
This prints time down to nanoseconds. Use Truncate method to truncate to minutes.

a-nogikh replied:
Added the truncation to seconds, I think they're important, too.

dvyukov commented on tools/syz-testbed/instance.go:
```diff
@@ -0,0 +1,129 @@
+			// The manager has exited.
+		case <-time.After(stopDelay):
+			// The manager did not exit - kill it.
+			log.Printf("[%s] instance did not exit itself, killing it", inst.Name)
```
I don't see where we are actually killing it. We need to wait for the process to finish on every return path.

dvyukov commented on tools/syz-testbed/instance.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func (inst *Instance) Stop() {
+	inst.stopChannel <- true
```
This can deadlock because the channel is unbuffered.
Consider 2 instances exit soon after start roughly at the same time. The main loop sees that one of the instances and calls Stop on the other instance. In this case send to stopChannel will hang forever because the instance.Run has already returned and nobody receives on the other end.
I think the simplest solution is to make the channel buffered.

dvyukov commented on tools/syz-testbed/instance.go:
```diff
@@ -0,0 +1,129 @@
+		return fmt.Errorf("[%s] stopped: %s", inst.Name, err)
+	case <-inst.stopChannel:
+		// TODO: handle other OSes?
+		cmd.Process.Signal(syscall.SIGTERM)
```
Use os.Interrupt, it's portable.
https://pkg.go.dev/os#Signal

a-nogikh commented:
@dvyukov PTAL

Changes since the last review:
* All the comments above fixed.
* Changed the config file - now `run_time` instead of `run_hours` that can accept strings like "15m".
* Add a new commit - `tools/syz-testbed: align checkout stats to the same time`.

dvyukov commented on tools/syz-testbed/instance.go:
```diff
@@ -0,0 +1,129 @@
+		case <-time.After(stopDelay):
+			// The manager did not exit - kill it.
+			log.Printf("[%s] instance did not exit itself, killing it", inst.Name)
+			cmd.Process.Kill()
```
Also add:
```
<-complete
```
to wait for the process.
